network/v1alpha1/types_dnsnameresolver.go

```diff
@@ -0,0 +1,119 @@
+type DNSNameResolver struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// metadata is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec is the specification of the desired behavior of the DNSNameResolver.
+	// +kubebuilder:validation:Required
+	Spec DNSNameResolverSpec `json:"spec"`
+	// status is the most recently observed status of the DNSNameResolver.
+	// +optional
+	Status DNSNameResolverStatus `json:"status,omitempty"`
+}
+
+// DNSNameResolverSpec is a desired state description of DNSNameResolver.
+type DNSNameResolverSpec struct {
+	// name is the DNS name for which the DNS name resolution information will be stored.
+	// For a regular DNS name, only the DNS name resolution information of the regular DNS
+	// name will be stored. For a wildcard DNS name, the DNS name resolution information
+	// of all the DNS names, that matches the wildcard DNS name, will be stored.
+	// For a wildcard DNS name, the '*' will match only one label. Additionally, only a single
+	// '*' can be used at the beginning of the wildcard DNS name. For example, '*.example.com.'
+	// will match 'sub1.example.com.' but won't match 'sub2.sub1.example.com.'
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Pattern=^(\*\.)?([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.$
+	// +kubebuilder:validation:MaxLength=254
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="spec.name is immutable"
+	Name string `json:"name"`
+}
+
+// DNSNameResolverStatus defines the observed status of DNSNameResolver.
+type DNSNameResolverStatus struct {
+	// resolvedNames contains a list of matching DNS names and their corresponding IP addresses
+	// along with TTL and last DNS lookup time.
+	// +listType=map
+	// +listMapKey=dnsName
+	// +patchMergeKey=dnsName
+	// +patchStrategy=merge
+	// +optional
+	ResolvedNames []DNSNameResolverStatusItem `json:"resolvedNames,omitempty" patchStrategy:"merge" patchMergeKey:"dnsName"`
+}
+
+// +kubebuilder:validation:Pattern=`^(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\/([0-9]|[1-2][0-9]|3[0-2])$|^s*((([0-9A-Fa-f]{1,4}:){7}(:|([0-9A-Fa-f]{1,4})))|(([0-9A-Fa-f]{1,4}:){6}:([0-9A-Fa-f]{1,4})?)|(([0-9A-Fa-f]{1,4}:){5}(((:[0-9A-Fa-f]{1,4}){0,1}):([0-9A-Fa-f]{1,4})?))|(([0-9A-Fa-f]{1,4}:){4}(((:[0-9A-Fa-f]{1,4}){0,2}):([0-9A-Fa-f]{1,4})?))|(([0-9A-Fa-f]{1,4}:){3}(((:[0-9A-Fa-f]{1,4}){0,3}):([0-9A-Fa-f]{1,4})?))|(([0-9A-Fa-f]{1,4}:){2}(((:[0-9A-Fa-f]{1,4}){0,4}):([0-9A-Fa-f]{1,4})?))|(([0-9A-Fa-f]{1,4}:){1}(((:[0-9A-Fa-f]{1,4}){0,5}):([0-9A-Fa-f]{1,4})?))|(:(:|((:[0-9A-Fa-f]{1,4}){1,7}))))(%.+)?s*/([0-9]|[1-9][0-9]|1[0-1][0-9]|12[0-8])$`
+// IPAddressStr is used for validation of an IP address.
+type IPAddressStr string
```

Comment:
I'd like to see some integration tests that exercise this validation please

Reply:
I have added tests for both IPv4 and IPv6 addresses.

Reply:
Another potential thing to try might be using a yaml-patch, not sure how exactly this works though in terms of validation errors, want to give it a try? https://github.com/openshift/api/blob/cd0541be0908c8204ae129594ca1cc3ccad51524/networkoperator/v1/001-egressrouter.crd.yaml-patch#L11-L15

Reply:
> @arkadeepsen Can you push to a separate branch and link me what you've tried? I'll take a look into it

I actually figured out the issue. The `ip` field is within `resolvedAddresses` map field, which is within `resolvedNames` map field. Both `resolvedAddresses` and `resolvedNames` need `maxItems` mentioned, without which the cost is exceeding budget. But I am not sure whether it'll be safe to have `maxItems` for these fields.

Reply:
> Another potential thing to try might be using a yaml-patch, not sure how exactly this works though in terms of validation errors, want to give it a try?
> 
> https://github.com/openshift/api/blob/cd0541be0908c8204ae129594ca1cc3ccad51524/networkoperator/v1/001-egressrouter.crd.yaml-patch#L11-L15

This looks interesting. I'll try this one out.

Reply:
> > @arkadeepsen Can you push to a separate branch and link me what you've tried? I'll take a look into it
> 
> I actually figured out the issue. The `ip` field is within `resolvedAddresses` map field, which is within `resolvedNames` map field. Both `resolvedAddresses` and `resolvedNames` need `maxItems` mentioned, without which the cost is exceeding budget. But I am not sure whether it'll be safe to have `maxItems` for these fields.

Ahh good find, the interesting part here is that I'm sure the cost of the regex is probably exceeding what the CEl validations are doing, that's a pretty complex regex, but, the apiserver doesn't seem to care about that limitation

Reply:
Format ipv4 and ipv6 don't support CIDR, so if you need CIDR format then this won't work, I think your regex accounts for CIDR format at the moment right?

Reply:
> Format ipv4 and ipv6 don't support CIDR, so if you need CIDR format then this won't work, I think your regex accounts for CIDR format at the moment right?

Here I need ipv4 or ipv6 format only as these are IP addresses associated with DNS names. However, cidr is also supported I think:

https://github.com/openshift/api/blob/cd0541be0908c8204ae129594ca1cc3ccad51524/config/v1/0000_10_config-operator_01_infrastructure-Default.crd.yaml-patch#L13-L15